pkg/crds/controllers/serverless/realtimeapi_controller_helpers.go

```diff
@@ -0,0 +1,571 @@
+		vs.Spec = desiredVirtualService.Spec
+
+		if err := ctrl.SetControllerReference(&api, &vs, r.Scheme); err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return op, err
+	}
+	return op, nil
+}
+
+func (r *RealtimeAPIReconciler) getEndpoint(ctx context.Context, api *serverless.RealtimeAPI) (string, error) {
+	req := client.ObjectKey{Namespace: consts.IstioNamespace, Name: "ingressgateway-apis"}
+	svc := kcore.Service{}
+	if err := r.Get(ctx, req, &svc); err != nil {
+		return "", err
+	}
+
+	ingress := svc.Status.LoadBalancer.Ingress
+	if ingress == nil || len(ingress) == 0 {
+		return "", nil
+	}
+
+	endpoint := urls.Join(
+		fmt.Sprintf("http://%s", svc.Status.LoadBalancer.Ingress[0].Hostname),
+		api.Spec.Networking.Endpoint,
+	)
+
+	return endpoint, nil
+}
+
+func (r *RealtimeAPIReconciler) desiredDeployment(api serverless.RealtimeAPI) kapps.Deployment {
+	containers, volumes := r.desiredContainers(api)
+	deploymentID, _, apiID := r.getOrCreateAPIIDs(api)
+
+	return *k8s.Deployment(&k8s.DeploymentSpec{
+		Name:           workloads.K8sName(api.Name),
+		Replicas:       api.Spec.Pod.Replicas,
+		MaxSurge:       pointer.String(api.Spec.UpdateStrategy.MaxSurge.String()),
+		MaxUnavailable: pointer.String(api.Spec.UpdateStrategy.MaxUnavailable.String()),
+		Labels: map[string]string{
+			"apiName":        api.Name,
+			"apiKind":        userconfig.RealtimeAPIKind.String(),
+			"apiID":          apiID,
+			"deploymentID":   deploymentID,
+			"cortex.dev/api": "true",
+		},
+		Annotations: r.generateAPIAnnotations(api),
+		Selector: map[string]string{
+			"apiName": api.Name,
+			"apiKind": userconfig.RealtimeAPIKind.String(),
+		},
+		PodSpec: k8s.PodSpec{
+			Labels: map[string]string{
+				"apiName":        api.Name,
+				"apiKind":        userconfig.RealtimeAPIKind.String(),
+				"deploymentID":   deploymentID,
+				"cortex.dev/api": "true",
+			},
+			Annotations: map[string]string{
+				"traffic.sidecar.istio.io/excludeOutboundIPRanges": "0.0.0.0/0",
+			},
+			K8sPodSpec: kcore.PodSpec{
+				RestartPolicy:                 kcore.RestartPolicyAlways,
+				TerminationGracePeriodSeconds: pointer.Int64(_terminationGracePeriodSeconds),
+				Containers:                    containers,
+				NodeSelector:                  workloads.NodeSelectors(),
+				Tolerations:                   workloads.GenerateResourceTolerations(),
+				Affinity:                      workloads.GenerateNodeAffinities(api.Spec.NodeGroups),
+				Volumes:                       volumes,
+				ServiceAccountName:            workloads.ServiceAccountName,
+			},
+		},
+	})
+}
+
+func (r *RealtimeAPIReconciler) desiredContainers(api serverless.RealtimeAPI) ([]kcore.Container, []kcore.Volume) {
+	containers, volumes := r.userContainers(api)
+	proxyContainer, proxyVolume := r.proxyContainer(api)
+
+	containers = append(containers, proxyContainer)
+	volumes = append(volumes, proxyVolume)
+
+	return containers, volumes
+}
```

Comment:
We should use the `workloads.RealtimeContainers(api)` function instead of the `userContainers` and `proxyContainer`. We don’t need duplicate code. Same code already also applies to the task/batch/async APIs, so we should use the `workloads` package. See how we’re doing it for the batch CRD for instance. You'll probably need a `serverless.RealtimeAPI` to `spec.API` translator. I believe that's how we do it for the batch CRD too.